src/lem/interpreter.rs

```diff
@@ -1,180 +1,260 @@
+/// information is used to generte the witness.
+#[derive(Clone)]
+pub struct Frame<F: LurkField> {
+    pub input: [Ptr<F>; 3],
+    pub output: [Ptr<F>; 3],
+    pub binds: HashMap<MetaPtr, Ptr<F>>,
+    pub visits: Visits<F>,
+}
+
+fn retrieve_many<F: LurkField>(
+    map: &HashMap<MetaPtr, Ptr<F>>,
+    args: &[MetaPtr],
+) -> Result<Vec<Ptr<F>>> {
+    args.iter()
+        .map(|mptr| {
+            let Some(ptr) = map.get(mptr).cloned() else {
+                bail!("{} not defined", mptr.name());
+            };
+            Ok(ptr)
+        })
+        .collect::<Result<Vec<_>>>()
 }
 
 impl LEM {
     /// Interprets a LEM using a stack of operations to be popped and executed.
-    /// It modifies a `Store` and assigns `Ptr`s to `MetaPtr`s as it goes.
-    pub fn run<F: LurkField>(
+    /// It modifies a `Store` and binds `MetaPtr`s to `Ptr`s as it goes. We also
+    /// want to collect data from visited slots.
+    fn run<F: LurkField>(
         &self,
         input: [Ptr<F>; 3],
         store: &mut Store<F>,
-    ) -> Result<Witness<F>> {
+        slots_info: &SlotsInfo,
+    ) -> Result<Frame<F>> {
         // key/val pairs on this map should never be overwritten
-        let mut ptrs = HashMap::default();
-        ptrs.insert(self.input[0].clone(), input[0]);
-        insert_into_ptrs(&mut ptrs, self.input[1].clone(), input[1])?;
-        insert_into_ptrs(&mut ptrs, self.input[2].clone(), input[2])?;
-        let mut output = None;
+        let mut binds = HashMap::default();
+        macro_rules! bind {
+            ( $mptr: expr, $ptr: expr ) => {
+                if binds.insert($mptr, $ptr).is_some() {
+                    bail!("{} already defined", $mptr);
+                }
+            };
+        }
+
+        for (i, name) in self.input.iter().enumerate() {
+            bind!(MetaPtr(name.clone()), input[i]);
+        }
+
+        let mut visits = Visits::default();
+
         let mut stack = vec![&self.lem_op];
         while let Some(op) = stack.pop() {
             match op {
                 LEMOP::Null(tgt, tag) => {
                     let tgt_ptr = Ptr::null(*tag);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
+                    bind!(tgt.clone(), tgt_ptr);
                 }
-                LEMOP::Hash2(tgt, tag, src) => {
-                    let src_ptr1 = src[0].get_ptr(&ptrs)?;
-                    let src_ptr2 = src[1].get_ptr(&ptrs)?;
-                    let tgt_ptr = store.intern_2_ptrs(*tag, src_ptr1, src_ptr2);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
+                LEMOP::Hash2(img, tag, preimg) => {
+                    let preimg_ptrs = retrieve_many(&binds, preimg)?;
+                    let tgt_ptr = store.intern_2_ptrs(*tag, preimg_ptrs[0], preimg_ptrs[1]);
+                    bind!(img.clone(), tgt_ptr);
+                    visits.insert(*slots_info.get_slot(op)?, preimg_ptrs);
```

Comment:
Can't we just take the next available slot? That was the simple design discussed long ago and assumed by my sketch PR.

Reply:
In other words, conceptually, I think this should just have something like:
```
self.slot_allocations.allocate(Hash2, tgt_ptr, preimg_ptrs);
```
The idea was that we just allocate slots in order, as encountered, when evaluating — and use the same strategy to choose slots when building the circuit.